File: algorithm/simulator/java/SortingSimulator.java
import java.util.*;

public class SortingSimulator {
    
    private int[] array;
    private int[] originalArray;
    private int comparisons;
    private int swaps;
    private int animationDelay;
    private boolean stepMode;
    private Scanner scanner;

    public SortingSimulator() {
        this.array = new int[0];
        this.originalArray = new int[0];
        this.comparisons = 0;
        this.swaps = 0;
        this.animationDelay = 300;
        this.stepMode = false;
        this.scanner = new Scanner(System.in);
    }

    public void setArray(int[] arr) {
        this.array = arr.clone();
        this.originalArray = arr.clone();
        this.comparisons = 0;
        this.swaps = 0;
    }

    public void randomArray(int size, int max) {
        Random rand = new Random();
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = rand.nextInt(max) + 1;
        }
        setArray(arr);
    }

    public void reset() {
        this.array = originalArray.clone();
        this.comparisons = 0;
        this.swaps = 0;
    }

    private void clearScreen() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
    }

    private void waitForInput() {
        if (stepMode) {
            System.out.print("\n[Press Enter to continue...]");
            scanner.nextLine();
        } else {
            try {
                Thread.sleep(animationDelay);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    private void displayState(String message, Set<Integer> comparing, Set<Integer> swapping, 
                               int pivot, Set<Integer> sorted) {
        clearScreen();
        System.out.println("═".repeat(70));
        System.out.println(" SORTING SIMULATOR ");
        System.out.println("═".repeat(70));
        System.out.println();
        
        if (message != null) {
            System.out.println("📍 " + message);
            System.out.println();
        }

        displayArray(comparing, swapping, pivot, sorted);
        System.out.println();
        displayStats();
        
        waitForInput();
    }

    private void displayArray(Set<Integer> comparing, Set<Integer> swapping, 
                               int pivot, Set<Integer> sorted) {
        if (array.length == 0) {
            System.out.println("  (empty array)");
            return;
        }

        int max = Arrays.stream(array).max().orElse(1);
        int height = 15;
        
        for (int h = height; h >= 1; h--) {
            StringBuilder row = new StringBuilder("  ");
            for (int i = 0; i < array.length; i++) {
                int barHeight = (int) Math.ceil((double) array[i] / max * height);
                if (barHeight >= h) {
                    if (sorted != null && sorted.contains(i)) {
                        row.append(" ██ ");
                    } else if (swapping != null && swapping.contains(i)) {
                        row.append(" ▓▓ ");
                    } else if (comparing != null && comparing.contains(i)) {
                        row.append(" ░░ ");
                    } else if (pivot == i) {
                        row.append(" ▒▒ ");
                    } else {
                        row.append(" ▌▌ ");
                    }
                } else {
                    row.append("    ");
                }
            }
            System.out.println(row);
        }
        
        StringBuilder indices = new StringBuilder("  ");
        StringBuilder values = new StringBuilder("  ");
        for (int i = 0; i < array.length; i++) {
            indices.append(String.format("[%2d]", i));
            values.append(String.format(" %2d ", array[i]));
        }
        System.out.println("─".repeat(4 + array.length * 4));
        System.out.println(values);
        System.out.println(indices);
        
        System.out.println();
        System.out.println("Legend: ▌▌=Normal  ░░=Comparing  ▓▓=Swapping  ▒▒=Pivot  ██=Sorted");
    }

    private void displayStats() {
        System.out.println("─".repeat(50));
        System.out.printf("Comparisons: %d | Swaps: %d | Array size: %d%n", 
                comparisons, swaps, array.length);
        System.out.println("─".repeat(50));
    }

    private void swap(int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
        swaps++;
    }

    public void quickSort() {
        System.out.println("\n" + "═".repeat(70));
        System.out.println(" QUICK SORT ");
        System.out.println("═".repeat(70));
        
        displayState("Starting Quick Sort", null, null, -1, null);
        quickSortHelper(0, array.length - 1, new HashSet<>());
        
        Set<Integer> allSorted = new HashSet<>();
        for (int i = 0; i < array.length; i++) allSorted.add(i);
        displayState("Quick Sort complete!", null, null, -1, allSorted);
    }

    private void quickSortHelper(int low, int high, Set<Integer> sorted) {
        if (low < high) {
            int pivotIdx = partition(low, high, sorted);
            sorted.add(pivotIdx);
            quickSortHelper(low, pivotIdx - 1, sorted);
            quickSortHelper(pivotIdx + 1, high, sorted);
        } else if (low == high) {
            sorted.add(low);
        }
    }

    private int partition(int low, int high, Set<Integer> sorted) {
        int pivot = array[high];
        displayState(String.format("Partition [%d..%d], pivot = %d", low, high, pivot), 
                     null, null, high, sorted);

        int i = low - 1;

        for (int j = low; j < high; j++) {
            comparisons++;
            Set<Integer> comparing = new HashSet<>(Arrays.asList(j, high));
            displayState(String.format("Compare arr[%d]=%d with pivot %d", j, array[j], pivot), 
                        comparing, null, high, sorted);

            if (array[j] < pivot) {
                i++;
                if (i != j) {
                    Set<Integer> swapping = new HashSet<>(Arrays.asList(i, j));
                    displayState(String.format("Swap arr[%d]=%d <-> arr[%d]=%d", i, array[i], j, array[j]),
                                null, swapping, high, sorted);
                    swap(i, j);
                }
            }
        }

        if (i + 1 != high) {
            Set<Integer> swapping = new HashSet<>(Arrays.asList(i + 1, high));
            displayState(String.format("Place pivot: Swap arr[%d]=%d <-> arr[%d]=%d", 
                        i + 1, array[i + 1], high, pivot), null, swapping, -1, sorted);
            swap(i + 1, high);
        }

        return i + 1;
    }

    public void mergeSort() {
        System.out.println("\n" + "═".repeat(70));
        System.out.println(" MERGE SORT ");
        System.out.println("═".repeat(70));
        
        displayState("Starting Merge Sort", null, null, -1, null);
        mergeSortHelper(0, array.length - 1);
        
        Set<Integer> allSorted = new HashSet<>();
        for (int i = 0; i < array.length; i++) allSorted.add(i);
        displayState("Merge Sort complete!", null, null, -1, allSorted);
    }

    private void mergeSortHelper(int left, int right) {
        if (left < right) {
            int mid = left + (right - left) / 2;
            mergeSortHelper(left, mid);
            mergeSortHelper(mid + 1, right);
            merge(left, mid, right);
        }
    }

    private void merge(int left, int mid, int right) {
        Set<Integer> merging = new HashSet<>();
        for (int i = left; i <= right; i++) merging.add(i);
        
        displayState(String.format("Merge [%d..%d] and [%d..%d]", left, mid, mid + 1, right),
                    merging, null, -1, null);

        int[] leftArr = Arrays.copyOfRange(array, left, mid + 1);
        int[] rightArr = Arrays.copyOfRange(array, mid + 1, right + 1);

        int i = 0, j = 0, k = left;

        while (i < leftArr.length && j < rightArr.length) {
            comparisons++;
            if (leftArr[i] <= rightArr[j]) {
                array[k] = leftArr[i];
                i++;
            } else {
                array[k] = rightArr[j];
                j++;
                swaps++;
            }
            k++;
            displayState(String.format("Merging: placed %d at index %d", array[k - 1], k - 1),
                        merging, null, -1, null);
        }

        while (i < leftArr.length) {
            array[k] = leftArr[i];
            i++;
            k++;
        }

        while (j < rightArr.length) {
            array[k] = rightArr[j];
            j++;
            k++;
        }
    }

    public void heapSort() {
        System.out.println("\n" + "═".repeat(70));
        System.out.println(" HEAP SORT ");
        System.out.println("═".repeat(70));
        
        displayState("Starting Heap Sort - Building max heap", null, null, -1, null);
        
        int n = array.length;
        
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(n, i, new HashSet<>());
        }
        displayState("Max heap built!", null, null, -1, null);

        Set<Integer> sorted = new HashSet<>();
        for (int i = n - 1; i > 0; i--) {
            Set<Integer> swapping = new HashSet<>(Arrays.asList(0, i));
            displayState(String.format("Swap root %d with arr[%d]=%d", array[0], i, array[i]),
                        null, swapping, -1, sorted);
            swap(0, i);
            sorted.add(i);
            heapify(i, 0, sorted);
        }
        sorted.add(0);
        
        displayState("Heap Sort complete!", null, null, -1, sorted);
    }

    private void heapify(int n, int i, Set<Integer> sorted) {
        int largest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < n) {
            comparisons++;
            Set<Integer> comparing = new HashSet<>(Arrays.asList(largest, left));
            displayState(String.format("Heapify: Compare arr[%d]=%d with arr[%d]=%d", 
                        largest, array[largest], left, array[left]), comparing, null, -1, sorted);
            if (array[left] > array[largest]) {
                largest = left;
            }
        }

        if (right < n) {
            comparisons++;
            Set<Integer> comparing = new HashSet<>(Arrays.asList(largest, right));
            displayState(String.format("Heapify: Compare arr[%d]=%d with arr[%d]=%d", 
                        largest, array[largest], right, array[right]), comparing, null, -1, sorted);
            if (array[right] > array[largest]) {
                largest = right;
            }
        }

        if (largest != i) {
            Set<Integer> swapping = new HashSet<>(Arrays.asList(i, largest));
            displayState(String.format("Heapify: Swap arr[%d]=%d <-> arr[%d]=%d", 
                        i, array[i], largest, array[largest]), null, swapping, -1, sorted);
            swap(i, largest);
            heapify(n, largest, sorted);
        }
    }

    public void runInteractiveMode() {
        System.out.println("═".repeat(70));
        System.out.println(" SORTING ALGORITHMS SIMULATOR ");
        System.out.println("═".repeat(70));
        System.out.println();
        System.out.println("Commands:");
        System.out.println("  set <values>    - Set array (comma-separated)");
        System.out.println("  random [n] [m]  - Random array of n elements (1 to m)");
        System.out.println("  quicksort       - Run Quick Sort");
        System.out.println("  mergesort       - Run Merge Sort");
        System.out.println("  heapsort        - Run Heap Sort");
        System.out.println("  display         - Show current array");
        System.out.println("  reset           - Reset to original array");
        System.out.println("  speed <ms>      - Set animation delay");
        System.out.println("  step            - Toggle step mode");
        System.out.println("  demo            - Run demo");
        System.out.println("  quit            - Exit");
        System.out.println();

        randomArray(10, 50);
        System.out.println("Generated random array: " + Arrays.toString(array));

        while (true) {
            System.out.print("sort> ");
            String input = scanner.nextLine().trim().toLowerCase();
            String[] parts = input.split("\\s+");

            if (parts.length == 0 || parts[0].isEmpty()) continue;

            String cmd = parts[0];

            try {
                switch (cmd) {
                    case "quit":
                    case "exit":
                        System.out.println("Goodbye!");
                        return;

                    case "set":
                        if (parts.length > 1) {
                            String[] values = input.substring(4).split("[,\\s]+");
                            int[] arr = Arrays.stream(values)
                                    .filter(s -> !s.isEmpty())
                                    .mapToInt(Integer::parseInt)
                                    .toArray();
                            setArray(arr);
                            System.out.println("Array set: " + Arrays.toString(array));
                        }
                        break;

                    case "random":
                        int n = parts.length > 1 ? Integer.parseInt(parts[1]) : 10;
                        int m = parts.length > 2 ? Integer.parseInt(parts[2]) : 50;
                        randomArray(n, m);
                        System.out.println("Random array: " + Arrays.toString(array));
                        break;

                    case "quicksort":
                        reset();
                        quickSort();
                        System.out.println("Sorted: " + Arrays.toString(array));
                        break;

                    case "mergesort":
                        reset();
                        mergeSort();
                        System.out.println("Sorted: " + Arrays.toString(array));
                        break;

                    case "heapsort":
                        reset();
                        heapSort();
                        System.out.println("Sorted: " + Arrays.toString(array));
                        break;

                    case "display":
                        displayState("Current array state", null, null, -1, null);
                        System.out.println("Array: " + Arrays.toString(array));
                        break;

                    case "reset":
                        reset();
                        System.out.println("Reset to: " + Arrays.toString(array));
                        break;

                    case "speed":
                        if (parts.length > 1) {
                            animationDelay = Integer.parseInt(parts[1]);
                            System.out.println("Delay set to " + animationDelay + "ms");
                        }
                        break;

                    case "step":
                        stepMode = !stepMode;
                        System.out.println("Step mode: " + (stepMode ? "ON" : "OFF"));
                        break;

                    case "demo":
                        runDemo();
                        break;

                    default:
                        System.out.println("Unknown command: " + cmd);
                }
            } catch (NumberFormatException e) {
                System.out.println("Error: Please enter valid integers");
            }
        }
    }

    public void runDemo() {
        System.out.println("\n" + "═".repeat(70));
        System.out.println(" DEMO - Sorting Algorithms Comparison ");
        System.out.println("═".repeat(70));
        System.out.println();
        
        setArray(new int[]{38, 27, 43, 3, 9, 82, 10});
        System.out.println("Demo array: " + Arrays.toString(array));
        
        System.out.println("\n1. Quick Sort");
        System.out.print("Press Enter to start...");
        scanner.nextLine();
        reset();
        quickSort();
        int qComp = comparisons, qSwap = swaps;
        
        System.out.println("\n2. Merge Sort");
        System.out.print("Press Enter to start...");
        scanner.nextLine();
        reset();
        mergeSort();
        int mComp = comparisons, mSwap = swaps;
        
        System.out.println("\n3. Heap Sort");
        System.out.print("Press Enter to start...");
        scanner.nextLine();
        reset();
        heapSort();
        int hComp = comparisons, hSwap = swaps;

        System.out.println("\n" + "═".repeat(70));
        System.out.println(" COMPARISON RESULTS ");
        System.out.println("═".repeat(70));
        System.out.printf("%-12s | Comparisons | Swaps%n", "Algorithm");
        System.out.println("─".repeat(35));
        System.out.printf("%-12s | %11d | %5d%n", "Quick Sort", qComp, qSwap);
        System.out.printf("%-12s | %11d | %5d%n", "Merge Sort", mComp, mSwap);
        System.out.printf("%-12s | %11d | %5d%n", "Heap Sort", hComp, hSwap);
    }

    public static void main(String[] args) {
        SortingSimulator sim = new SortingSimulator();

        if (args.length > 0 && args[0].equals("--demo")) {
            sim.runDemo();
        } else {
            sim.runInteractiveMode();
        }
    }
}

File: code/simulator/sort/python/sorting_simulator.py
#!/usr/bin/env python3
"""
Sorting Algorithms Interactive Simulator
Quick Sort, Merge Sort, Heap Sort 과정을 시각화합니다.
"""

import sys
import os
import time
import random


class SortingSimulator:
    def __init__(self, array=None):
        self.array = array if array else []
        self.original_array = self.array.copy()
        self.comparisons = 0
        self.swaps = 0
        self.step_mode = False
        self.animation_delay = 0.3
        self.history = []

    def clear_screen(self):
        os.system("clear" if os.name != "nt" else "cls")

    def render_array(
        self, highlight_indices=None, pivot_idx=None, sorted_indices=None, label=""
    ):
        highlight_indices = highlight_indices or []
        sorted_indices = sorted_indices or set()

        print("\n" + "─" * 70)
        if label:
            print(f" {label} ".center(70))
            print("─" * 70)

        max_val = max(self.array) if self.array else 1
        bar_width = 50

        for i, val in enumerate(self.array):
            bar_length = int((val / max_val) * bar_width)

            if i == pivot_idx:
                marker = "P→"
                bar_char = "█"
            elif i in highlight_indices:
                marker = "→ "
                bar_char = "▓"
            elif i in sorted_indices:
                marker = "✓ "
                bar_char = "░"
            else:
                marker = "  "
                bar_char = "▒"

            bar = bar_char * bar_length
            print(f"{marker}[{i:2d}] {val:4d} │{bar}")

        print("─" * 70)
        print(f"Comparisons: {self.comparisons} | Swaps: {self.swaps}")

    def render_merge_state(self, left, right, merged, label=""):
        print("\n" + "=" * 70)
        print(f" {label} ".center(70))
        print("=" * 70)

        print(f"\n  Left:   {left}")
        print(f"  Right:  {right}")
        print(f"  Merged: {merged}")

    def quick_sort(self, low=None, high=None, depth=0):
        if low is None:
            low = 0
        if high is None:
            high = len(self.array) - 1
            self.clear_screen()
            print("=" * 70)
            print(" QUICK SORT - Starting ".center(70))
            print("=" * 70)
            self.render_array(label="Initial Array")
            if self.step_mode:
                input("\nPress Enter to start...")

        if low < high:
            pivot_idx = self._partition(low, high, depth)

            self.quick_sort(low, pivot_idx - 1, depth + 1)
            self.quick_sort(pivot_idx + 1, high, depth + 1)

        if depth == 0:
            self.clear_screen()
            print("=" * 70)
            print(" QUICK SORT - Complete! ".center(70))
            print("=" * 70)
            self.render_array(
                sorted_indices=set(range(len(self.array))), label="Sorted Array"
            )

    def _partition(self, low, high, depth):
        pivot = self.array[high]
        indent = "  " * depth

        self.clear_screen()
        print("=" * 70)
        print(f" PARTITION (depth={depth}) ".center(70))
        print("=" * 70)
        print(f"\n{indent}Range: [{low}...{high}]")
        print(f"{indent}Pivot: {pivot} (index {high})")

        self.render_array(
            highlight_indices=list(range(low, high)),
            pivot_idx=high,
            label=f"Partitioning around pivot {pivot}",
        )

        if self.step_mode:
            input("\nPress Enter to continue...")
        else:
            time.sleep(self.animation_delay)

        i = low - 1

        for j in range(low, high):
            self.comparisons += 1

            if self.array[j] <= pivot:
                i += 1
                if i != j:
                    self.array[i], self.array[j] = self.array[j], self.array[i]
                    self.swaps += 1

                    self.clear_screen()
                    print("=" * 70)
                    print(
                        f" SWAP: arr[{i}]={self.array[j]} ↔ arr[{j}]={self.array[i]} ".center(
                            70
                        )
                    )
                    print("=" * 70)
                    self.render_array(highlight_indices=[i, j], pivot_idx=high)

                    if self.step_mode:
                        input("\nPress Enter to continue...")
                    else:
                        time.sleep(self.animation_delay * 0.5)

        self.array[i + 1], self.array[high] = self.array[high], self.array[i + 1]
        self.swaps += 1

        self.clear_screen()
        print("=" * 70)
        print(f" PIVOT PLACED at index {i + 1} ".center(70))
        print("=" * 70)
        self.render_array(pivot_idx=i + 1, label="Partition Complete")

        if self.step_mode:
            input("\nPress Enter to continue...")
        else:
            time.sleep(self.animation_delay)

        return i + 1

    def merge_sort(self, arr=None, depth=0, start_idx=0):
        if arr is None:
            arr = self.array.copy()
            self.clear_screen()
            print("=" * 70)
            print(" MERGE SORT - Starting ".center(70))
            print("=" * 70)
            self.render_array(label="Initial Array")
            if self.step_mode:
                input("\nPress Enter to start...")

        if len(arr) <= 1:
            return arr

        mid = len(arr) // 2
        indent = "  " * depth

        self.clear_screen()
        print("=" * 70)
        print(f" DIVIDE (depth={depth}) ".center(70))
        print("=" * 70)
        print(f"\n{indent}Splitting: {arr}")
        print(f"{indent}  Left:  {arr[:mid]}")
        print(f"{indent}  Right: {arr[mid:]}")

        if self.step_mode:
            input("\nPress Enter to continue...")
        else:
            time.sleep(self.animation_delay)

        left = self.merge_sort(arr[:mid], depth + 1, start_idx)
        right = self.merge_sort(arr[mid:], depth + 1, start_idx + mid)

        merged = self._merge(left, right, depth)

        for i, val in enumerate(merged):
            self.array[start_idx + i] = val

        if depth == 0:
            self.clear_screen()
            print("=" * 70)
            print(" MERGE SORT - Complete! ".center(70))
            print("=" * 70)
            self.render_array(
                sorted_indices=set(range(len(self.array))), label="Sorted Array"
            )

        return merged

    def _merge(self, left, right, depth):
        indent = "  " * depth
        merged = []
        i = j = 0

        self.clear_screen()
        print("=" * 70)
        print(f" MERGE (depth={depth}) ".center(70))
        print("=" * 70)
        print(f"\n{indent}Merging:")
        print(f"{indent}  Left:  {left}")
        print(f"{indent}  Right: {right}")

        while i < len(left) and j < len(right):
            self.comparisons += 1

            if left[i] <= right[j]:
                merged.append(left[i])
                i += 1
            else:
                merged.append(right[j])
                j += 1

        merged.extend(left[i:])
        merged.extend(right[j:])

        print(f"{indent}  Result: {merged}")

        if self.step_mode:
            input("\nPress Enter to continue...")
        else:
            time.sleep(self.animation_delay)

        return merged

    def heap_sort(self):
        n = len(self.array)

        self.clear_screen()
        print("=" * 70)
        print(" HEAP SORT - Starting ".center(70))
        print("=" * 70)
        self.render_array(label="Initial Array")

        if self.step_mode:
            input("\nPress Enter to start...")

        for i in range(n // 2 - 1, -1, -1):
            self._heapify(n, i, "Building Max Heap")

        self.clear_screen()
        print("=" * 70)
        print(" MAX HEAP BUILT ".center(70))
        print("=" * 70)
        self.render_array(label="Max Heap")
        self._render_heap_tree()

        if self.step_mode:
            input("\nPress Enter to continue...")
        else:
            time.sleep(self.animation_delay)

        sorted_set = set()
        for i in range(n - 1, 0, -1):
            self.array[0], self.array[i] = self.array[i], self.array[0]
            self.swaps += 1
            sorted_set.add(i)

            self.clear_screen()
            print("=" * 70)
            print(f" EXTRACT MAX: {self.array[i]} → index {i} ".center(70))
            print("=" * 70)
            self.render_array(highlight_indices=[0, i], sorted_indices=sorted_set)

            if self.step_mode:
                input("\nPress Enter to continue...")
            else:
                time.sleep(self.animation_delay * 0.5)

            self._heapify(i, 0, "Re-heapify")

        sorted_set.add(0)
        self.clear_screen()
        print("=" * 70)
        print(" HEAP SORT - Complete! ".center(70))
        print("=" * 70)
        self.render_array(sorted_indices=sorted_set, label="Sorted Array")

    def _heapify(self, heap_size, root_idx, label=""):
        largest = root_idx
        left = 2 * root_idx + 1
        right = 2 * root_idx + 2

        if left < heap_size:
            self.comparisons += 1
            if self.array[left] > self.array[largest]:
                largest = left

        if right < heap_size:
            self.comparisons += 1
            if self.array[right] > self.array[largest]:
                largest = right

        if largest != root_idx:
            self.array[root_idx], self.array[largest] = (
                self.array[largest],
                self.array[root_idx],
            )
            self.swaps += 1
            self._heapify(heap_size, largest, label)

    def _render_heap_tree(self):
        if not self.array:
            return

        print("\n  Heap Tree Visualization:")
        n = len(self.array)
        level = 0
        idx = 0

        while idx < n:
            level_size = 2**level
            indent = " " * (2 ** (4 - level))
            spacing = " " * (2 ** (5 - level))

            values = []
            for _ in range(level_size):
                if idx < n:
                    values.append(f"{self.array[idx]:3d}")
                    idx += 1
                else:
                    break

            print(indent + spacing.join(values))
            level += 1

    def reset(self):
        self.array = self.original_array.copy()
        self.comparisons = 0
        self.swaps = 0
        self.history = []

    def set_array(self, arr):
        self.array = arr.copy()
        self.original_array = arr.copy()
        self.comparisons = 0
        self.swaps = 0

    def randomize(self, size=10, max_val=50):
        self.array = [random.randint(1, max_val) for _ in range(size)]
        self.original_array = self.array.copy()
        self.comparisons = 0
        self.swaps = 0


def print_help():
    print("""
╔══════════════════════════════════════════════════════════════════════╗
║                   SORTING SIMULATOR - HELP                           ║
╠══════════════════════════════════════════════════════════════════════╣
║ Commands:                                                            ║
║   set <values>    - Set array (e.g., set 5 3 8 1 9)                  ║
║   random [n] [m]  - Generate random array (n elements, max m)        ║
║   quicksort       - Run Quick Sort                                   ║
║   mergesort       - Run Merge Sort                                   ║
║   heapsort        - Run Heap Sort                                    ║
║   display         - Show current array                               ║
║   reset           - Reset to original array                          ║
║   speed <mode>    - Set speed (instant/fast/normal/slow)             ║
║   step            - Toggle step-by-step mode                         ║
║   demo            - Run demo with all algorithms                     ║
║   help            - Show this help                                   ║
║   quit            - Exit simulator                                   ║
╚══════════════════════════════════════════════════════════════════════╝
    """)


def run_demo_mode():
    print("\n" + "=" * 70)
    print(" DEMO MODE - Sorting Algorithms ".center(70))
    print("=" * 70)

    demo_array = [64, 34, 25, 12, 22, 11, 90]

    print(f"\nDemo array: {demo_array}")
    print("\nThis demo will show:")
    print("1. Quick Sort")
    print("2. Merge Sort")
    print("3. Heap Sort")

    input("\nPress Enter to start with Quick Sort...")

    sim = SortingSimulator(demo_array.copy())
    sim.animation_delay = 0.4
    sim.quick_sort()

    print(f"\nQuick Sort completed!")
    print(f"Comparisons: {sim.comparisons}, Swaps: {sim.swaps}")
    input("\nPress Enter to continue with Merge Sort...")

    sim.set_array(demo_array)
    sim.merge_sort()

    print(f"\nMerge Sort completed!")
    print(f"Comparisons: {sim.comparisons}")
    input("\nPress Enter to continue with Heap Sort...")

    sim.set_array(demo_array)
    sim.heap_sort()

    print(f"\nHeap Sort completed!")
    print(f"Comparisons: {sim.comparisons}, Swaps: {sim.swaps}")

    print("\n" + "=" * 70)
    print(" DEMO COMPLETED ".center(70))
    print("=" * 70)


def run_interactive_mode():
    print("=" * 70)
    print(" SORTING ALGORITHMS SIMULATOR ".center(70))
    print("=" * 70)

    sim = SortingSimulator([64, 34, 25, 12, 22, 11, 90])
    print_help()

    while True:
        try:
            cmd = input("\nsort> ").strip().lower()
        except EOFError:
            break
        except KeyboardInterrupt:
            print("\n")
            break

        if not cmd:
            continue

        parts = cmd.split()
        command = parts[0]

        if command == "quit" or command == "exit":
            print("Goodbye!")
            break

        elif command == "help":
            print_help()

        elif command == "set":
            if len(parts) < 2:
                print("Usage: set <value1> <value2> ...")
                continue
            try:
                arr = [int(x) for x in parts[1:]]
                sim.set_array(arr)
                print(f"✅ Array set: {arr}")
            except ValueError:
                print("Error: Please enter valid integers")

        elif command == "random":
            size = int(parts[1]) if len(parts) > 1 else 10
            max_val = int(parts[2]) if len(parts) > 2 else 50
            sim.randomize(size, max_val)
            print(f"✅ Random array: {sim.array}")

        elif command == "quicksort":
            sim.quick_sort()

        elif command == "mergesort":
            sim.merge_sort()

        elif command == "heapsort":
            sim.heap_sort()

        elif command == "display":
            sim.render_array(label="Current Array")

        elif command == "reset":
            sim.reset()
            print(f"✅ Reset to: {sim.array}")

        elif command == "speed":
            if len(parts) < 2:
                print("Usage: speed <instant/fast/normal/slow>")
                continue
            mode = parts[1]
            speeds = {"instant": 0, "fast": 0.1, "normal": 0.3, "slow": 0.8}
            if mode in speeds:
                sim.animation_delay = speeds[mode]
                print(f"✅ Speed set to {mode}")
            else:
                print("Unknown speed mode")

        elif command == "step":
            sim.step_mode = not sim.step_mode
            print(f"✅ Step-by-step mode: {'ON' if sim.step_mode else 'OFF'}")

        elif command == "demo":
            run_demo_mode()

        else:
            print(f"Unknown command: {command}. Type 'help' for available commands.")


if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "--demo":
        run_demo_mode()
    else:
        run_interactive_mode()
